File: XCTestBootstrap/Reporters/FBLogicReporterAdapter.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBLogicReporterAdapter.h"
#import <XCTestBootstrap/FBXCTestReporter.h>
#import <XCTestBootstrap/FBXCTestLogger.h>

#import "FBXCTestConstants.h"

@interface FBLogicReporterAdapter ()

@property (nonatomic, readonly) id<FBXCTestReporter> reporter;
@property (nonatomic, readonly) FBXCTestLogger *logger;

@end

@implementation FBLogicReporterAdapter

- (instancetype)initWithReporter:(id<FBXCTestReporter>)reporter logger:(nullable id<FBControlCoreLogger>)logger
{
  self = [self init];
  if (!self) {
    return nil;
  }
  _reporter = reporter;
  _logger = [logger withName:@"FBLogicReporterAdapter"];

  return self;
}


- (void)didBeginExecutingTestPlan
{
  [self.reporter didBeginExecutingTestPlan];
}

- (void)didFinishExecutingTestPlan
{
  [self.reporter didFinishExecutingTestPlan];
  [self.reporter processUnderTestDidExit];
}

- (void)processWaitingForDebuggerWithProcessIdentifier:(pid_t)pid
{
  [self.reporter processWaitingForDebuggerWithProcessIdentifier:pid];
}

- (void)testHadOutput:(NSString *)output
{
  [self.reporter testHadOutput:output];
}

- (void)handleEventJSONData:(NSData *)data
{
  if (data.length == 0) {
    [self.logger log:@"Received zero-length JSON data"];
    return;
  }
  NSError *error;
  NSDictionary<NSString *, id> *JSONEvent = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
  if (![JSONEvent isKindOfClass:[NSDictionary class]]) {
    [self.logger logFormat:@"Received invalid JSON: '%@' %@", [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding], error];
    return;
  }

  NSString *eventName = JSONEvent[kReporter_Event_Key];
  id<FBXCTestReporter> reporter = self.reporter;
  if ([eventName isEqualToString:kReporter_Events_BeginTestSuite]) {
    NSString *suiteName = JSONEvent[kReporter_BeginTestSuite_SuiteKey];
    id startTime = JSONEvent[kReporter_TimestampKey];
    if ([startTime isKindOfClass:NSNumber.class]) {
      [reporter testSuite:suiteName didStartAt:((NSNumber *)startTime).stringValue];
    } else if ([startTime isKindOfClass:NSString.class]) {
      [reporter testSuite:suiteName didStartAt:((NSString *)startTime)];
    } else {
      NSAssert(NO, @"Unknown type of obj. This will likely cause crash in runtime because of swift signature mismatch");
    }

  } else if ([eventName isEqualToString:kReporter_Events_BeginTest]) {
    NSString *className = JSONEvent[kReporter_BeginTest_ClassNameKey];
    NSString *methodName = JSONEvent[kReporter_BeginTest_MethodNameKey];
    [reporter testCaseDidStartForTestClass:className method:methodName];
  } else if ([eventName isEqualToString:kReporter_Events_EndTest]) {
    [self handleEndTest:JSONEvent data:data];
  } else if ([eventName isEqualToString:kReporter_Events_EndTestSuite]) {
    NSDate *finishDate = [NSDate dateWithTimeIntervalSince1970:[JSONEvent[kReporter_TimestampKey] doubleValue]];
    NSInteger unexpected = [JSONEvent[kReporter_EndTestSuite_UnexpectedExceptionCountKey] integerValue];
    FBTestManagerResultSummary *summary = [[FBTestManagerResultSummary alloc]
      initWithTestSuite:JSONEvent[kReporter_EndTestSuite_SuiteKey]
      finishTime:finishDate
      runCount:[JSONEvent[kReporter_EndTestSuite_TestCaseCountKey] integerValue]
      failureCount:[JSONEvent[kReporter_EndTestSuite_TotalFailureCountKey] integerValue]
      unexpected:unexpected
      testDuration:[JSONEvent[kReporter_EndTestSuite_TestDurationKey] doubleValue]
      totalDuration:[JSONEvent[kReporter_EndTest_TotalDurationKey] doubleValue]];
    [reporter finishedWithSummary:summary];
  } else {
    [self.logger logFormat:@"[%@] Unhandled event JSON: %@", NSStringFromClass(self.class), JSONEvent];
    //We don't know how to handle it, but an upstream reporter might.
    NSString *stringEvent = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    [reporter handleExternalEvent:stringEvent];
  }
}

- (void)handleEndTest:(NSDictionary<NSString *, id> *)JSONEvent data:(NSData *)data
{
  id<FBXCTestReporter> reporter = self.reporter;
  NSString *testClass = JSONEvent[kReporter_EndTest_ClassNameKey];
  NSString *testName = JSONEvent[kReporter_EndTest_MethodNameKey];
  NSString *result = JSONEvent[kReporter_EndTest_ResultKey];
  NSTimeInterval duration = [JSONEvent[kReporter_EndTest_TotalDurationKey] doubleValue];

  if ([result isEqualToString:kReporter_EndTest_ResultValueSuccess]) {
    [reporter testCaseDidFinishForTestClass:testClass method:testName withStatus:FBTestReportStatusPassed duration:duration logs:nil];
  } else if ([[NSSet setWithArray:@[kReporter_EndTest_ResultValueFailure, kReporter_EndTest_ResultValueError]] containsObject:result]) {
    [self reportTestFailureForTestClass:testClass testName:testName endTestEvent:JSONEvent];
    [reporter testCaseDidFinishForTestClass:testClass method:testName withStatus:FBTestReportStatusFailed duration:duration logs:nil];
  } else {
    // We don't know how to handle it, but an upstream reporter might.
    NSString *stringEvent = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    [reporter handleExternalEvent:stringEvent];
  }
}

- (void)reportTestFailureForTestClass:(NSString *)testClass testName:(NSString *)testName endTestEvent:(NSDictionary *)JSONEvent
{
  NSArray<NSDictionary *> *exceptionDicts = JSONEvent[kReporter_EndTest_ExceptionsKey];
    NSMutableArray<FBExceptionInfo *> *parsedExceptions = [NSMutableArray new];

    for (NSDictionary *exceptionDict in exceptionDicts) {
        NSString *message = exceptionDict[kReporter_EndTest_Exception_ReasonKey];
        NSString *file = exceptionDict[kReporter_EndTest_Exception_FilePathInProjectKey];
        NSUInteger line = [exceptionDict[kReporter_EndTest_Exception_LineNumberKey] unsignedIntegerValue];
        FBExceptionInfo *exception = [[FBExceptionInfo alloc]initWithMessage:message file:file line:line];
        [parsedExceptions addObject:exception];
    }

  [self.reporter testCaseDidFailForTestClass:testClass method:testName exceptions:[parsedExceptions copy]];
}

- (void)didCrashDuringTest:(NSError *)error
{
  if ([self.reporter respondsToSelector:@selector(didCrashDuringTest:)]) {
    [self.reporter didCrashDuringTest:error];
  }
  [self.reporter processUnderTestDidExit];
}

@end

File: XCTestBootstrap/Reporters/FBLogicReporterAdapter.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

@objc public final class FBLogicReporterAdapter: NSObject, FBLogicXCTestReporter {

  private let reporter: FBXCTestReporter
  private let logger: FBControlCoreLogger?

  @objc public init(reporter: FBXCTestReporter, logger: FBControlCoreLogger?) {
    self.reporter = reporter
    self.logger = logger?.withName("FBLogicReporterAdapter") as (any FBControlCoreLogger)?
    super.init()
  }

  // MARK: FBLogicXCTestReporter

  @objc public func didBeginExecutingTestPlan() {
    reporter.didBeginExecutingTestPlan()
  }

  @objc public func didFinishExecutingTestPlan() {
    reporter.didFinishExecutingTestPlan()
    reporter.processUnderTestDidExit()
  }

  @objc public func processWaitingForDebugger(withProcessIdentifier pid: pid_t) {
    reporter.processWaitingForDebugger(withProcessIdentifier: pid)
  }

  @objc public func testHadOutput(_ output: String) {
    reporter.testHadOutput(output)
  }

  @objc public func handleEventJSONData(_ data: Data) {
    if data.isEmpty {
      logger?.log("Received zero-length JSON data")
      return
    }
    guard let jsonEvent = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
      logger?.log("Received invalid JSON: '\(String(data: data, encoding: .utf8) ?? "")'")
      return
    }

    let eventName = jsonEvent["event"] as? String
    if eventName == "begin-test-suite" {
      let suiteName = jsonEvent["suite"] as? String ?? ""
      let startTime = jsonEvent["timestamp"]
      if let startTime = startTime as? NSNumber {
        reporter.testSuite(suiteName, didStartAt: startTime.stringValue)
      } else if let startTime = startTime as? String {
        reporter.testSuite(suiteName, didStartAt: startTime)
      } else {
        assertionFailure("Unknown type of obj. This will likely cause crash in runtime because of swift signature mismatch")
      }
    } else if eventName == "begin-test" {
      let className = jsonEvent["className"] as? String ?? ""
      let methodName = jsonEvent["methodName"] as? String ?? ""
      reporter.testCaseDidStart(forTestClass: className, method: methodName)
    } else if eventName == "end-test" {
      handleEndTest(jsonEvent, data: data)
    } else if eventName == "end-test-suite" {
      let finishDate = Date(timeIntervalSince1970: (jsonEvent["timestamp"] as? NSNumber)?.doubleValue ?? 0)
      let unexpected = (jsonEvent["unexpectedExceptionCount"] as? NSNumber)?.intValue ?? 0
      let summary = FBTestManagerResultSummary(
        testSuite: jsonEvent["suite"] as? String ?? "",
        finishTime: finishDate,
        runCount: (jsonEvent["testCaseCount"] as? NSNumber)?.intValue ?? 0,
        failureCount: (jsonEvent["totalFailureCount"] as? NSNumber)?.intValue ?? 0,
        unexpected: unexpected,
        testDuration: (jsonEvent["testDuration"] as? NSNumber)?.doubleValue ?? 0,
        totalDuration: (jsonEvent["totalDuration"] as? NSNumber)?.doubleValue ?? 0
      )
      reporter.finished(with: summary)
    } else {
      logger?.log("[\(String(describing: type(of: self)))] Unhandled event JSON: \(jsonEvent)")
      // We don't know how to handle it, but an upstream reporter might.
      let stringEvent = String(data: data, encoding: .utf8) ?? ""
      reporter.handleExternalEvent(stringEvent)
    }
  }

  @objc public func didCrashDuringTest(_ error: Error) {
    if reporter.responds(to: #selector(FBXCTestReporter.didCrashDuringTest(_:))) {
      reporter.didCrashDuringTest(error as NSError)
    }
    reporter.processUnderTestDidExit()
  }

  // MARK: Private

  private func handleEndTest(_ jsonEvent: [String: Any], data: Data) {
    let testClass = jsonEvent["className"] as? String ?? ""
    let testName = jsonEvent["methodName"] as? String ?? ""
    let result = jsonEvent["result"] as? String ?? ""
    let duration = (jsonEvent["totalDuration"] as? NSNumber)?.doubleValue ?? 0

    if result == "success" {
      reporter.testCaseDidFinish(forTestClass: testClass, method: testName, with: .passed, duration: duration, logs: nil)
    } else if result == "failure" || result == "error" {
      reportTestFailure(forTestClass: testClass, testName: testName, endTestEvent: jsonEvent)
      reporter.testCaseDidFinish(forTestClass: testClass, method: testName, with: .failed, duration: duration, logs: nil)
    } else {
      // We don't know how to handle it, but an upstream reporter might.
      let stringEvent = String(data: data, encoding: .utf8) ?? ""
      reporter.handleExternalEvent(stringEvent)
    }
  }

  private func reportTestFailure(forTestClass testClass: String, testName: String, endTestEvent jsonEvent: [String: Any]) {
    guard let exceptionDicts = jsonEvent["exceptions"] as? [[String: Any]] else { return }
    var parsedExceptions: [FBExceptionInfo] = []

    for exceptionDict in exceptionDicts {
      let message = exceptionDict["reason"] as? String ?? ""
      let file = exceptionDict["filePathInProject"] as? String
      let line = (exceptionDict["lineNumber"] as? NSNumber)?.uintValue ?? 0
      let exception = FBExceptionInfo(message: message, file: file, line: line)
      parsedExceptions.append(exception)
    }

    reporter.testCaseDidFail(forTestClass: testClass, method: testName, exceptions: parsedExceptions)
  }
}
